File: 01_Exercise_Basic_Blade_Design/Solution/Exercise01_Solution.py
import numpy as np
import matplotlib.pyplot as plt

# a) Find R, lambda_D, and z
R = 63  # [m]
lambda_D = 7.55  # [-]
z = 3  # [-]

# b) Find optimal AoA and cL
# Import Airfoil file
AirfoilFile = r'IntrodutionToWindTurbineAerodynamics\01_Exercise_Basic_Blade_Design\Solution\IEA-15-240-RWT_AeroDyn15_Polar_36.dat'
AirfoilData = np.loadtxt(AirfoilFile, skiprows=54)

AoA = AirfoilData[:, 0]
Cl = AirfoilData[:, 1]
Cd = AirfoilData[:, 2]
Cm = AirfoilData[:, 3]

# Calculate lift-to-drag ratio
epsilon = Cl / Cd

# Find maximum value
MaxIdx = np.argmax(epsilon)
MaxVal = epsilon[MaxIdx]

# Plot Lift and Drag
plt.figure()
plt.grid(True)
plt.box(True)
plt.plot(AoA, Cl, label='Lift')
plt.plot(AoA, Cd, label='Drag')
plt.xlabel(r'\alpha_{A} [deg]')
plt.ylabel(r'c_{L}, c_{D} [-]')
plt.legend()
plt.show()

# Plot lift-to-drag ratio
plt.figure()
plt.grid(True)
plt.box(True)
plt.plot(AoA, epsilon)
plt.xlabel(r'\alpha_{A} [deg]')
plt.ylabel(r'\epsilon [-]')
plt.show()

# c) Find twist and chord
# Import AeroDyn file
AeroFile = r'IntrodutionToWindTurbineAerodynamics\01_Exercise_Basic_Blade_Design\Material\NRELOffshrBsline5MW_AeroDyn_Equil_noTwr.dat'
NodeData = np.loadtxt(AeroFile, skiprows=26, usecols=(0, 1, 2, 3, 4))

RNodes = NodeData[:, 0]
AeroTwst = NodeData[:, 1]
DRNodes = NodeData[:, 2]
Chord = NodeData[:, 3]
NFoil = NodeData[:, 4]

# Missing design parameters
alpha_A = AoA[MaxIdx]  # [deg]
c_L = Cl[MaxIdx]  # [-]

# Calculate twist angle
r = RNodes
beta = np.degrees(np.arctan((2/3) * R / lambda_D / r)) - alpha_A
c = (1 / z) * (8/9) * (2 * np.pi * R / c_L) * (1 / (lambda_D * np.sqrt((lambda_D * r / R) ** 2 + 4/9)))

# d) Compare to NREL results
plt.figure()
plt.grid(True)
plt.box(True)
plt.plot(RNodes, beta, '.-', label='Betz')
plt.plot(RNodes, AeroTwst, 'o-', label='NREL')
plt.xlabel('r [m]')
plt.ylabel(r'\beta [deg]')
plt.title('Comparison Twist Angle')
plt.legend()
plt.show()

plt.figure()
plt.grid(True)
plt.box(True)
plt.plot(RNodes, c, '.-', label='Betz')
plt.plot(RNodes, Chord, 'o-', label='NREL')
plt.xlabel('r [m]')
plt.ylabel('c [m]')
plt.title('Comparison Chord Length')
plt.legend()
plt.show()

File: 01_Exercise_Basic_Blade_Design/Solution/Exercise01_Solution.m
% -----------------------------
% Solution to Exercise 01 of Course "Introduction to Wind Turbine Aerodynamics".
% ----------------------------------
clearvars;close all;clc

%% a) find R, lambda_D and z
% design parameters: please adjust
R               = 63;               % [m]
lambda_D        = 7.55;           	% [-]
z               = 3;                % [-]

%% b) find optimal AoA and cL
% import Airfoil file
AirfoilFile  	= 'NACA64_A17.dat'; 
fid            	= fopen(AirfoilFile);
AirfoilData   	= textscan(fid,'%f %f %f %f','HeaderLines',14);
fclose(fid);
BEM.AoA         = AirfoilData{1};
BEM.Cl          = AirfoilData{2};
BEM.Cd          = AirfoilData{3};
BEM.Cm          = AirfoilData{4};

% calculate lift-to-drag-ratio
epsilon         = BEM.Cl./BEM.Cd;

% find maximum value
[MaxVal,MaxIdx] = max(epsilon);

% plot
figure
box on;grid on;hold on
plot(BEM.AoA,[BEM.Cl,BEM.Cd])
xlabel('\alpha_{{A}} [deg]')
ylabel('c_{L}, c_{D} [-]')
legend({'Lift','Drag'})

figure
box on;grid on;hold on
plot(BEM.AoA,epsilon)
xlabel('\alpha_{{A}} [deg]')
ylabel('\epsilon [-]')

%% c) find twist and chord
% import AeroDyn file
AeroFile        = 'NRELOffshrBsline5MW_AeroDyn_Equil_noTwr.dat';
fid             = fopen(AeroFile);
NodeData        = textscan(fid,'%f %f %f %f %d %s','HeaderLines',26);
fclose(fid);
BEM.RNodes  	= NodeData{1};
BEM.AeroTwst 	= NodeData{2};
BEM.DRNodes     = NodeData{3};
BEM.Chord       = NodeData{4};
BEM.NFoil       = NodeData{5};

% missing design parameters: please adjust
alpha_A         = BEM.AoA(MaxIdx);	% [deg]
c_L             = BEM.Cl(MaxIdx);   % [-]

% calculate twist angle: please adjust
r               = BEM.RNodes;
beta            = atand(2/3*R/lambda_D./r)-alpha_A;
c               = 1/z*8/9*2*pi*R/c_L*1./(lambda_D*sqrt((lambda_D*r/R).^2+4/9));

%% d) compare to NREL results
figure
box on;grid on;hold on
plot(BEM.RNodes,beta,'.-')
plot(BEM.RNodes,BEM.AeroTwst,'o-')
xlabel('r [m]')
ylabel('\beta [deg]')
title('Comparison Twist Angle')
legend('Betz','NREL')

figure
box on;grid on;hold on
plot(BEM.RNodes,c,'.-')
plot(BEM.RNodes,BEM.Chord,'o-')
xlabel('r [m]')
ylabel('c [m]')
title('Comparison Chord Length')
legend('Betz','NREL')